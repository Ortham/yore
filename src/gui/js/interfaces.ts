export interface Coordinates {
  latitude: number;
  longitude: number;
}

export interface Photo {
  path: string;
  location?: Location; // eslint-disable-line no-restricted-globals
  error?: string;
  height?: number;
  width?: number;
  loaded?: boolean;
}

interface Location {
  Existing?: Coordinates;
  Suggested?: [Coordinates, LocationAccuracy];
}

export interface LocationAccuracy {
  meters: number;
  seconds: number;
}

export interface LocationResponse {
  location?: Location; // eslint-disable-line no-restricted-globals
  error?: string;
}
