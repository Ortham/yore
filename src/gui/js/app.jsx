import 'babel-polyfill';
import 'whatwg-fetch';
import React from 'react';
import ReactDOM from 'react-dom';
import Page from './page';
import { getPhotos, getRootPath } from './requests';

function getInitialState() {
  const state = {
    rootPath: undefined,
    filterPhotos: false,
    currentPhoto: undefined,
    photos: []
  };

  return getRootPath()
    .then(responseBody => {
      state.rootPath = responseBody.rootPath;

      return getPhotos();
    })
    .then(photos => {
      state.photos = photos;

      return state;
    });
}

getInitialState().then(state => {
  ReactDOM.render(
    <Page rootPath={state.rootPath} photos={state.photos} />,
    document.getElementById('root')
  );
});
