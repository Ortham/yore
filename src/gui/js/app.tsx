import '@babel/polyfill';
import * as React from 'react';
import * as ReactDOM from 'react-dom';
import 'whatwg-fetch';
import { Page, PageProps } from './page';
import {
  getInterpolate,
  getLocationHistoryPath,
  getPhotos,
  getRootPath,
  getLocations
} from './requests';
import { updatePhotoLocations } from './photo';

function getInitialState() {
  const state: PageProps = {
    interpolate: false,
    locationHistoryPath: undefined,
    photos: [],
    rootPath: undefined
  };

  return getRootPath()
    .then(responseBody => {
      state.rootPath = responseBody.rootPath || '';

      return getPhotos();
    })
    .then(photos =>
      getLocations(0, photos.length).then(locations =>
        updatePhotoLocations(photos, locations)
      )
    )
    .then(photos => {
      state.photos = photos;

      return getInterpolate();
    })
    .then(responseBody => {
      state.interpolate = responseBody.interpolate;

      return getLocationHistoryPath();
    })
    .then(responseBody => {
      state.locationHistoryPath = responseBody.locationHistoryPath || '';

      return state;
    });
}

getInitialState().then(state => {
  ReactDOM.render(<Page {...state} />, document.getElementById('root'));
});
