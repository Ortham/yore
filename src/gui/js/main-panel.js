import React from 'react';
import MapArea from './map-area';
import {locationDescription, hasSuggestion} from './photo';

export default function MainPanel(props) {
    if (props.photo) {
        return (
            <main>
                <section>
                    <img src={props.photo.src} />
                    <MapArea photo={props.photo} />
                </section>
                <footer>
                    {locationDescription(props.photo)}
                    <br />
                    <button
                        disabled={!hasSuggestion(props.photo)}
                        onClick={props.handleSuggestionApply} >
                        Apply
                    </button>
                    <button
                        disabled={!hasSuggestion(props.photo)}
                        onClick={props.handleSuggestionDiscard} >
                        Discard
                    </button>
                </footer>
            </main>
        );
    } else {
        return (
            <main>
                <section>No photo selected.</section>
                <footer>
                    <br />
                    <button disabled>Apply</button>
                    <button disabled>Discard</button>
                </footer>
            </main>
        );
    }
}
