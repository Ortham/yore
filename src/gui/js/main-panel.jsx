import React from 'react';
import PropTypes from 'prop-types';
import MapArea from './map-area';
import { locationDescription, hasSuggestion } from './photo';

export default function MainPanel(props) {
  if (props.photo) {
    return (
      <div id="main">
        <section>
          <img src={props.photo.src} alt="Selected" />
          <MapArea photo={props.photo} />
        </section>
        <footer>
          <div>{locationDescription(props.photo)}</div>
          <div>
            <button
              disabled={!hasSuggestion(props.photo)}
              onClick={props.handleSuggestionApply}
            >
              Apply
            </button>
            <button
              disabled={!hasSuggestion(props.photo)}
              onClick={props.handleSuggestionDiscard}
            >
              Discard
            </button>
          </div>
        </footer>
      </div>
    );
  }
  return (
    <div id="main">
      <section>No photo selected.</section>
      <footer>
        <div />
        <div>
          <button disabled>Apply</button>
          <button disabled>Discard</button>
        </div>
      </footer>
    </div>
  );
}

MainPanel.propTypes = {
  photo: PropTypes.shape({
    src: PropTypes.string
  }),
  handleSuggestionApply: PropTypes.func.isRequired,
  handleSuggestionDiscard: PropTypes.func.isRequired
};

MainPanel.defaultProps = {
  photo: undefined
};
