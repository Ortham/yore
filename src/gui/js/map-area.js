import React from 'react';
import GoogleMapReact from 'google-map-react';
import FaMapMarker from 'react-icons/lib/fa/map-marker';
import {googleMapsCoordinates} from './photo';

function MapMarker(props) {
    return <FaMapMarker lat={props.lat} lng={props.lng} style={{height: '24px', width: '24px', color: 'crimson'}} />;
}

export default function MapArea(props) {
    if (props.photo && props.photo.location) {
        const coord = googleMapsCoordinates(props.photo);
        return (
            <GoogleMapReact center={coord} zoom={5}>
                <MapMarker lat={coord.lat} lng={coord.lng} />
            </GoogleMapReact>
        );
    } else {
        return <GoogleMapReact center={[0, 0]} zoom={5} />;
    }
}
