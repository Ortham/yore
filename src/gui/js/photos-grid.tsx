import * as React from 'react';
import Pig from 'pig.js/src/pig.js';
import { Photo } from './interfaces';
import PigPhotoThumbnail from './pig-photo-thumbnail';

export interface PhotosGridProps {
  photos: Photo[];
  currentPhoto?: Photo;
  handlePhotoSelect: (photo: Photo) => void;
}

export class PhotosGrid extends React.Component<PhotosGridProps, {}> {
  private pig: any;

  public componentDidMount() {
    const imageData = this.getPigImageData();

    const photosList = document.getElementById('photosList');

    const options = {
      containerId: 'pig',
      scroller: photosList,
      imageType: PigPhotoThumbnail,

      // Width is 10*height to avoid getting a thumbnail constrained by its
      // width.
      urlForSize: (filename: string, height: number) =>
        `/thumbnail?path=${encodeURIComponent(filename)}&maxWidth=${10 *
          height}&maxHeight=${height}`
    };

    this.pig = new Pig(imageData, options).enable();
    // Immediately call update adjust layout to account for scroll bar.
    this.pig.update(imageData);
  }

  public componentWillUnmount() {
    this.pig.disable();
  }

  public componentDidUpdate(/* prevProps, prevState, snapshot */) {
    if (this.pig) {
      this.pig.update(this.getPigImageData());
    }
  }

  public render() {
    return (
      <nav id="photosList">
        <div id="pig" style={{ width: '100%' }} />
      </nav>
    );
  }

  private getPigImageData() {
    return this.props.photos.map(photo => ({
      filename: photo.path,
      aspectRatio: photo.width / photo.height,
      photo,
      isSelected: photo === this.props.currentPhoto,
      handlePhotoSelect: this.props.handlePhotoSelect
    }));
  }
}
