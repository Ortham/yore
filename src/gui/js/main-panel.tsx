import * as React from 'react';
import { Photo } from './interfaces';
import { MapArea } from './map-area';
import { hasSuggestion, locationDescription } from './photo';

export interface MainPanelProps {
  photo?: Photo;
  handleSuggestionApply: () => Promise<void>;
  handleSuggestionDiscard: () => void;
}

export function MainPanel(props: MainPanelProps) {
  if (props.photo) {
    return (
      <div id="main">
        <section>
          <img src={props.photo.src} alt="Selected" />
          <MapArea photo={props.photo} />
        </section>
        <footer>
          <div>{locationDescription(props.photo)}</div>
          <div>
            <button
              disabled={!hasSuggestion(props.photo)}
              onClick={props.handleSuggestionApply}
            >
              Apply
            </button>
            <button
              disabled={!hasSuggestion(props.photo)}
              onClick={props.handleSuggestionDiscard}
            >
              Discard
            </button>
          </div>
        </footer>
      </div>
    );
  }
  return (
    <div id="main">
      <section>No photo selected.</section>
      <footer>
        <div />
        <div>
          <button disabled>Apply</button>
          <button disabled>Discard</button>
        </div>
      </footer>
    </div>
  );
}
