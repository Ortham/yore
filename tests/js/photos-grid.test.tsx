import * as React from 'react';
import * as renderer from 'react-test-renderer';
import { PhotosGrid } from '../../src/gui/js/photos-grid';

describe('PhotosGrid', () => {
  let mockGetAndStoreLocations: jest.Mock;
  let mockHandlePhotoSelect: jest.Mock;
  let filteredPhotosGrid: renderer.ReactTestRenderer;

  const photos = [
    {
      height: 5,
      width: 10,
      loaded: false,
      path: 'path',
      src: 'source'
    },
    {
      height: 5,
      width: 10,
      loaded: true,
      path: 'path',
      src: 'source'
    }
  ];

  beforeAll(() => {
    mockGetAndStoreLocations = jest.fn();
    mockHandlePhotoSelect = jest.fn();

    filteredPhotosGrid = renderer.create(
      <PhotosGrid
        photos={photos}
        getAndStoreLocations={mockGetAndStoreLocations}
        handlePhotoSelect={mockHandlePhotoSelect}
      />
    );
  });

  beforeEach(() => {
    mockGetAndStoreLocations.mockClear();
    mockHandlePhotoSelect.mockClear();
  });

  test('renders a grid of photo thumbnails and a filter checkbox and label', () => {
    const photosGrid = renderer
      .create(
        <PhotosGrid
          photos={photos}
          getAndStoreLocations={mockGetAndStoreLocations}
          handlePhotoSelect={mockHandlePhotoSelect}
        />
      )
      .toJSON();
    expect(photosGrid).toMatchSnapshot();
  });

  test('onSectionRendered calls onRowsRendered with the correct indices', () => {
    const photosGrid = filteredPhotosGrid.root.instance;
    photosGrid.columnCount = 5;

    photosGrid.onRowsRendered = jest.fn();

    photosGrid.onSectionRendered({
      columnStartIndex: 1,
      columnStopIndex: 3,
      rowStartIndex: 2,
      rowStopIndex: 4
    });

    expect(photosGrid.onRowsRendered.mock.calls.length).toBe(1);
    expect(photosGrid.onRowsRendered.mock.calls[0].length).toBe(1);
    expect(photosGrid.onRowsRendered.mock.calls[0][0]).toEqual({
      startIndex: 11,
      stopIndex: 23
    });
  });

  test('cellRenderer calls rowRenderer with the correct photo index', () => {
    const photosGrid = filteredPhotosGrid.root.instance;
    photosGrid.columnCount = 5;

    const originalRowRenderer = photosGrid.rowRenderer;
    photosGrid.rowRenderer = jest.fn();

    photosGrid.cellRenderer({
      columnIndex: 1,
      key: 5,
      rowIndex: 3,
      style: { height: 'auto' }
    });

    expect(photosGrid.rowRenderer.mock.calls.length).toBe(1);
    expect(photosGrid.rowRenderer.mock.calls[0].length).toBe(1);
    expect(photosGrid.rowRenderer.mock.calls[0][0]).toEqual({
      index: 16,
      key: 5,
      style: { height: 'auto' }
    });

    photosGrid.rowRenderer = originalRowRenderer;
  });

  test('rowRenderer returns a PhotoThumbnail for the photo at the given index', () => {
    const photosGrid = filteredPhotosGrid.root.instance;

    const photoThumbnail = photosGrid.rowRenderer({
      index: 0,
      key: 1,
      style: { color: 'black' }
    });
    expect(photoThumbnail).toMatchSnapshot();
  });

  test('rowRender should set the PhotoThumbnail handleSelect to call the handlePhotoSelect callback', () => {
    const photosGrid = filteredPhotosGrid.root.instance;

    const photoThumbnail = photosGrid.rowRenderer({
      index: 0,
      key: 1,
      style: { color: 'black' }
    });

    photoThumbnail.props.handleSelect();

    expect(mockHandlePhotoSelect.mock.calls.length).toBe(1);
    expect(mockHandlePhotoSelect.mock.calls[0].length).toBe(1);
    expect(mockHandlePhotoSelect.mock.calls[0][0]).toBe(photos[0]);
  });

  test('rowHeight scales the height of the photo at the given index to match a width of 272', () => {
    const photosGrid = filteredPhotosGrid.root.instance;
    photosGrid.columnCount = 2;

    expect(photosGrid.rowHeight({ index: 0 })).toBe(75);
  });

  test('isRowLoaded returns false if the photo at the given index has loaded = false', () => {
    const photosGrid = filteredPhotosGrid.root.instance;

    expect(photosGrid.isRowLoaded({ index: 0 })).toBe(false);
  });

  test('isRowLoaded returns true if the photo at the given index has loaded = true', () => {
    const photosGrid = filteredPhotosGrid.root.instance;

    expect(photosGrid.isRowLoaded({ index: 1 })).toBe(true);
  });

  test('loadMoreRows should call the getAndStoreLocations callback', () => {
    const photosGrid = filteredPhotosGrid.root.instance;

    photosGrid.loadMoreRows({ startIndex: 0, stopIndex: 1 });
    expect(mockGetAndStoreLocations.mock.calls.length).toBe(1);
    expect(mockGetAndStoreLocations.mock.calls[0]).toEqual([0, 1]);
  });

  test('forceUpdate should force the grid to update and reset the loaded rows cache', () => {
    const photosGrid = filteredPhotosGrid.root.instance;

    photosGrid.grid.recomputeGridSize = jest.fn();
    photosGrid.loader.resetLoadMoreRowsCache = jest.fn();

    photosGrid.forceUpdate();

    expect(photosGrid.grid.recomputeGridSize.mock.calls.length).toBe(1);
    expect(photosGrid.loader.resetLoadMoreRowsCache.mock.calls.length).toBe(1);
    expect(photosGrid.loader.resetLoadMoreRowsCache.mock.calls[0]).toEqual([
      true
    ]);
  });
});
