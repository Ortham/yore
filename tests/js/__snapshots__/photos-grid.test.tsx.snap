// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PhotosGrid renders a grid of photo thumbnails and a filter checkbox and label 1`] = `
<nav
  id="photosList"
>
  <div
    style={
      Object {
        "height": 0,
        "overflow": "visible",
        "width": 0,
      }
    }
  >
    <div
      aria-label="grid"
      aria-readonly={true}
      className="ReactVirtualized__Grid"
      onScroll={[Function]}
      role="grid"
      style={
        Object {
          "WebkitOverflowScrolling": "touch",
          "boxSizing": "border-box",
          "direction": "ltr",
          "height": 0,
          "overflowX": "hidden",
          "overflowY": "auto",
          "position": "relative",
          "width": 0,
          "willChange": "transform",
        }
      }
      tabIndex={0}
    />
  </div>
</nav>
`;

exports[`PhotosGrid rowRenderer returns a PhotoThumbnail for the photo at the given index 1`] = `
<PhotoThumbnail
  handleSelect={[Function]}
  isSelected={false}
  photo={
    Object {
      "height": 5,
      "loaded": false,
      "path": "path",
      "src": "source",
      "width": 10,
    }
  }
  style={
    Object {
      "color": "black",
    }
  }
/>
`;
